WalletWasabi/Tor/Socks5/TorPoolItemManager.cs

```diff
@@ -0,0 +1,158 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using WalletWasabi.Logging;
+using WalletWasabi.Tor.Socks5.Pool;
+
+namespace WalletWasabi.Tor.Socks5
+{
+	/// <summary>
+	/// Associates <see cref="IPoolItem"/> with <see cref="PoolItemState"/> and manages pool items re-usability.
+	/// </summary>
+	/// <remarks>The class is thread-safe.</remarks>
+	public class TorPoolItemManager: IDisposable
+	{
+		public TorPoolItemManager(int maxPoolItemsPerHost)
+		{
+			HostBuckets = new Dictionary<string, List<IPoolItem>>();
+			MaxPoolItemsPerHost = maxPoolItemsPerHost;
+		}
+
+		/// <summary>Provider of exclusive access to <see cref="HostBuckets"/>.</summary>
+		private object HostBucketsLock { get; } = new object();
+
+		/// <summary>Key is always a URI host. Value is a list of pool items that can connect to the URI host.</summary>
+		/// <remarks>All access to this object must be guarded by <see cref="HostBucketsLock"/>.</remarks>
+		private Dictionary<string, List<IPoolItem>> HostBuckets { get; }
+
+		public int MaxPoolItemsPerHost { get; }
+
+		private bool _disposedValue;
+
+		/// <summary>
+		/// Adds <paramref name="poolItem"/> to <see cref="HostBuckets"/>.
+		/// </summary>
+		/// <param name="host">URI's host.</param>
+		/// <param name="poolItem">Pool item to add.</param>
+		/// <returns><c>true</c> when the new item was added, <c>false</c> otherwise.</returns>
+		public bool AddPoolItem(string host, IPoolItem poolItem)
+		{
+			lock (HostBucketsLock)
+			{
+				// Get list of connections for given host.
+				List<IPoolItem> hostItems = GetHostListNoLock(host);
+
+				if (hostItems.Count < MaxPoolItemsPerHost)
+				{
+					hostItems.Add(poolItem);
+					return true;
+				}
+				else
+				{
+					return false;
+				}
+			}
+		}
+
+		/// <summary>
+		/// Gets all <see cref="IPoolItem"/>s which are related to URI <paramref name="host"/>.
+		/// </summary>
+		/// <param name="host">A <see cref="Uri.Host"/> value is expected.</param>
+		public List<IPoolItem> GetItemsCopy(string host)
+		{
+			lock (HostBucketsLock)
+			{
+				return new List<IPoolItem>(GetHostListNoLock(host));
+			}
+		}
+
+		/// <summary>
+		/// Gets reserved pool item to use, if any.
+		/// </summary>
+		/// <param name="host">URI's host value.</param>
+		/// <param name="isolateStream"><c>true</c> if a new Tor circuit is required for this HTTP request.</param>
+		/// <returns>Whether a new pool item can be added to <see cref="TorPoolItemManager"/> and reserved pool item to use, if any.</returns>
+		public (bool canBeAdded, IPoolItem? poolItem) GetPoolItem(string host, bool isolateStream)
```

Comment:
Right, it probably used to be an `async` method where it is not allowed to have `out` parameters.

> Why `canBeAdded` returned here, it makes no sense to me.

There is `MaxPoolItemsPerHost` limit which sets maximum number of concurrent connections to a single URI host. This may or may not be a useful feature.